yjit/src/options.rs

```diff
@@ -0,0 +1,124 @@
+use std::ffi::CStr;
+
+// Command-line options
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+#[repr(C)]
+pub struct Options {
+    // Size of the executable memory block to allocate in MiB
+    pub exec_mem_size : usize,
+
+    // Number of method calls after which to start generating code
+    // Threshold==1 means compile on first execution
+    pub call_threshold : usize,
+
+    // Generate versions greedily until the limit is hit
+    pub greedy_versioning : bool,
+
+    // Disable the propagation of type information
+    pub no_type_prop : bool,
+
+    // Maximum number of versions per block
+    // 1 means always create generic versions
+    pub max_versions : usize,
+
+    // Capture and print out stats
+    pub gen_stats : bool,
+
+    /// Dump compiled and executed instructions for debugging
+    pub dump_insns: bool,
+
+    /// Verify context objects (debug mode only)
+    pub verify_ctx: bool,
+
+    /// Whether or not to assume a global constant state (and therefore
+    /// invalidating code whenever any constant changes) versus assuming
+    /// constant name components (and therefore invalidating code whenever a
+    /// matching name component changes)
+    pub global_constant_state: bool,
+}
+
+// Initialize the options to default values
+pub static mut OPTIONS: Options = Options {
+    exec_mem_size : 256,
+    call_threshold : 10,
+    greedy_versioning : false,
+    no_type_prop : false,
+    max_versions : 4,
+    gen_stats : false,
+    dump_insns: false,
+    verify_ctx: false,
+    global_constant_state: false,
+};
+
+/// Macro to get an option value by name
+macro_rules! get_option {
+    // Unsafe is ok here because options are initialized
+    // once before any Ruby code executes
+    ($option_name:ident) => {
+        unsafe {
+            OPTIONS.$option_name
+        }
+    };
+}
+pub(crate) use get_option;
+
+/// Expected to receive what comes after the third dash in "--yjit-*".
+/// Empty string means user passed only "--yjit". C code rejects when
+/// they pass exact "--yjit-".
+pub fn parse_option(str_ptr: *const std::os::raw::c_char) -> Option<()>
+{
+    let c_str: &CStr = unsafe { CStr::from_ptr(str_ptr) };
```

Comment:
i suppose to preclude accidental reuse of `c_str` without null checking done by `opt_str` you could simplify the first two lines as follows:

`let opt_str: &str = unsafe { CStr::from_ptr(str_ptr) }.to_str().ok()?;`

Reply:
Sorry, to be clear, I think they meant we need to check for null pointers here, not null termination. `CStr::from_ptr` must be called with a non null pointer (perhaps `parse_option` could accept a `NonNull<c_char>`?)

Reply:
ahh, good catch! yeah, that makes sense.

Reply:
Oh yeah sorry I wasn't clear, Lucretiel is correct in that I meant a null pointer being input rather than a null sentinel. I would also be careful with just raw Nonnull since this is a pub function that is called directly from C FFI. 

What I would consider "idiomatic" would be having the C FFI code do the conversions to Rust happy types. This is just a quick example but this is approximately how I would do it. But obviously the choice is yours.
```rust
use std::os::raw::c_char;
use std::ptr::NonNull;
use std::ffi::CStr;

// I would have it as an optional non-null since this is pub so this is no guarantee that someone will provide a null.
// I wouldn't change unless this conversion turns out to be a performance bottleneck. If it does, then putting an 
// `.unwrap_unchecked()` would be the move cause it still expresses intent.
pub extern "C" fn rb_yjit_parse_option(str_ptr: Option<NonNull<raw::c_char>>) -> bool
{
    match str_ptr {
        Some(str_ptr) => {
            let new_str = unsafe { 
                CStr::from_ptr(str_ptr.as_ptr()).to_str()
            };
            match new_str {
                Ok(rust_str) => parse_option(rust_str).is_some(),
                Err(_) => false
            }
        },
        None => false
    }
}
```